### ISSUE TITLE
Run validation every N batches

### ISSUE THREAD

**Issue by mpetteno** [keras-team/keras#19562]
Hi everyone.

This is not really an issue but more of a help request, I'm sorry if this maybe is not the right place to ask.
I'm trying to train a model over an entire dataset loaded with Tensorflow `tf.data` API: since the dataset contains a lot of samples, my idea is to run the training for only one epoch but launch the validation every n batches in order to be able to save the best model according to the validation losss. Is this possible in Keras?

Looking at the code in `fit` it seems to me that the evaluation is run only at the end of each epoch and even if I set `steps_per_epoch` then the next epoch the elements will be taken from the beginning again, thus I'm not using the whole dataset. Is that correct or am I missing something?

I tried to build a custom callback to run the validation and it works, but then I have to manuall handle the history and other callbacks that runs only at the end of each epoch.

Thanks for your help.

**Comment by fchollet** [comment]
> Looking at the code in fit it seems to me that the evaluation is run only at the end of each epoch and even if I set steps_per_epoch then the next epoch the elements will be taken from the beginning again, thus I'm not using the whole dataset. Is that correct or am I missing something?

Can you check that in practice with an example? I expect that you will continue drawing from the same dataset. It needs to be able to generate `steps_per_epoch * epochs` batches. You may need to call `.repeat()` on it to achieve it.

**Comment by mpetteno** [comment]
Yes the point is that I don't want to use `.repeat()` beacause my dataset is already big enough. I want to be able to run a single epoch (using the whole train dataset) but run the validation every N train steps and save the model if there's an improvement of the validation loss. Now, for the last part I know I can use `ModelCheckpoint` callback but I'm not sure if I can use the built-in `fit` method or not to achieve the first requirement.

**Comment by fchollet** [comment]
Fair enough, doing this in your own callback is a good solution. Lets you customize it to do whatever you want.

**Comment by mpetteno** [comment]
But how can I update the history in order to keep track also of the training loss? Because at the moment the history for that will be saved at the of each epoch that in my case it's only one

**Comment by fchollet** [comment]
Just make your callback create & update its own metrics/loss dict?

**Comment by mpetteno** [comment]
Ok, not sure if I got it but I'll try that thanks. I was hoping that there was a way to do this by using the standard loop in the `fit` method.  At the end what I want it's like running N epochs but reading the training dataset from a certain index and not from the beginning at each epoch.

**Comment by mpetteno** [comment]
The issue can be closed because it is not really an issue. Setting the seed and `deterministic=True` while loading the dataset with `tf.data` API helped in understing how Keras works and to achieve the desired result.
Keras support validation every N batches natively by using `steps_per_epoch` just be sure to call `.repeat()` on the dataset if necessary, as stated by @fchollet above.

The custom callback is not necessary.

Thanks for helping to clarify this.

**Comment by mpetteno** [comment]
The issue can be closed because it is not really an issue. Setting the seed and `deterministic=True` while loading the dataset with `tf.data` API helped in understing how Keras works and to achieve the desired result.
Keras support validation every N batches natively by using `steps_per_epoch` just be sure to call `.repeat()` on the dataset if necessary, as stated by @fchollet above.

**Comment by raevillena** [comment]
you may want to check validation frequency. I don't know when was that added but it is there now since this latest version.

**Comment by dulearnaux** [comment]
> The issue can be closed because it is not really an issue. Setting the seed and `deterministic=True` while loading the dataset with `tf.data` API helped in understing how Keras works and to achieve the desired result. Keras support validation every N batches natively by using `steps_per_epoch` just be sure to call `.repeat()` on the dataset if necessary, as stated by [@fchollet](https://github.com/fchollet) above.
> 
> The custom callback is not necessary.
> 
> Thanks for your help in clarifying this.

Wouldn't setting `steps_per_epoch` to a smaller number just re-use the begining of a `tf.data.Dataset` after `steps_per_epoch` steps are completed (i.e. each time a psuedo `epoch` is completed)? But the desired behaviour is to use the next unused part of a `tf.data.Dataset` each time `epoch` turns over, untill the whole `tf.data.Dataset` is consumed, then start at the begining again?
